### DISCUSSION TITLE
Brilliant idea!

### DISCUSSION THREAD

**Discussion by mityakoval** [Homebrew/discussions#3085] [Everyday usage]
Who got the idea to update everything that has ever been installed on one's machine before installing a single package, and then repairing everything afterwards? And also enabling that by default?

**Comment by SMillerDev** [comment]
The volunteers who got really tired of users that don't update their system only to complain things aren't working after.

https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

**Comment by mityakoval** [comment]
Shouldn't it be at least prompted?
I would assume that most homebrew users are not installing stuff everyday.
This probably happens occasionally when something new is being set up. And when it is needed to install something quickly, which should be a very straight forward process, it turns out I need to wait an hour so that everything I have ever installed is being updated and cleaned up. And I don't even need any of those packages at the moment. And it's not clear if interrupting this process is safe. 

And how is that even a justification, that something doesn't work in someone else's environment?

Why did this never happen to me when using `apt`?

**Reply by SMillerDev** [reply]
> Shouldn't it be at least prompted?

Nothing in Homebrew works with prompts to optimize compatibility with scripts.

> I would assume that most homebrew users are not installing stuff everyday.
> This probably happens occasionally when something new is being set up. And when it is needed to install something quickly, which should be a very straight forward process, it turns out I need to wait an hour so that everything I have ever installed is being updated and cleaned up. And I don't even need any of those packages at the moment. 

Well, not keeping your system up to date is not really a situation we optimize for.

> And it's not clear if interrupting this process is safe.

Generally I wouldn't recommend it.

> And how is that even a justification, that something doesn't work in someone else's environment?

Because you'd be even unhappier if all the software you install is broken until you do an update.

> Why did this never happen to me when using apt?

Because Debian, and by extenstion apt, have fundamentally different ideas about when to update software. See https://en.wikipedia.org/wiki/Rolling_release for the differences. Additionally, apt does actually download and install updates if they are required. Maybe you simply update your Debian/Ubuntu system automatically/more often.

**Reply by tlk** [reply]
A while back `brew` worked like `apt` with a separate step to update[1] and another step to install or upgrade a software package and its dependencies. Personally, I prefer this mode and have re-enabled it by setting the `HOMEBREW_NO_AUTO_UPDATE=1` environment variable.

These days however, `brew install` will occasionally sneak in an automatic update step. On sunny days this may go unnoticed without any issues, but on rainy days this may cause cascading upgrades resulting in b0rked packages, etc.

I found that this change in behaviour was introduced in 2016:
* https://github.com/Homebrew/legacy-homebrew/issues/38818
* https://github.com/Homebrew/brew/issues/288

(trigger-warning: not a pleasant discussion to read, IMO)


[1] `apt update` / `brew update`:  to download the most recent knowledge about available packages.
